File: sdk/python/geotrace-py/examples/channels.py
#!/usr/bin/env python3
"""Write a .gtd file with ad-hoc sensor channels, then read them back.

A channel is a named time series sampled at its own rate, correlated with the
nav track by timestamp. It can be scalar (an inclinometer angle) or a vector
whose components share one sample clock (an accelerometer's x/y/z axes).

The three channels below cover the three ways to declare a unit: a recognized
label string, a Unit catalog constant, and ChannelUnit.custom for a label
outside the catalog, whose values stay dimensionless in queries.
"""

from __future__ import annotations

import tempfile
from datetime import UTC, datetime, timedelta
from pathlib import Path

from geotrace_sdk import Channel, ChannelUnit, NavFile, NavFileBuilder, NavFix, Unit

START = datetime(2024, 6, 1, 8, 0, 0, tzinfo=UTC)


def main() -> None:
    times = [START + timedelta(seconds=i) for i in range(3)]

    builder = NavFileBuilder().with_title("Channel tour")
    builder.add(NavFix(lat=51.5074, lon=-0.1278, gps_time=START))
    builder.add(
        Channel(
            "incline",
            times,
            [1.0, 1.5, 2.0],
            unit="deg",
            description="boom inclinometer",
        )
    )
    builder.add(
        Channel(
            "accel",
            times,
            # Row-major: three samples of (x, y, z).
            [
                0.0,
                200.0,
                980.0,
                100.0,
                200.0,
                980.0,
                200.0,
                200.0,
                980.0,
            ],
            unit=Unit.MG,
            components=["x", "y", "z"],
        )
    )
    builder.add(
        Channel(
            "quality",
            times,
            [80.0, 81.0, 82.0],
            unit=ChannelUnit.custom("vendor score"),
        )
    )

    out = Path(tempfile.gettempdir()) / "geotrace_channels.gtd"
    builder.finish().write_to_file(out)
    try:
        nav_file = NavFile.open(out)
        print(f"{len(nav_file.channels)} channels:")
        for channel in nav_file.channels:
            line = f"  {channel.name:<10} {len(channel.times)} samples"
            if channel.unit:
                line += f" [{channel.unit}]"
            if channel.is_vector:
                line += f" components: {' '.join(channel.components)}"
            print(line)
    finally:
        out.unlink()


if __name__ == "__main__":
    main()

File: sdk/rust/geotrace-sdk/examples/channels.rs
//! Write a `.gtd` file with ad-hoc sensor channels, then read them back.
//!
//! A [`Channel`] is a named time series sampled at its own rate, correlated
//! with the nav track by timestamp - not resampled onto the fixes. It can be
//! scalar (an inclinometer angle) or a vector whose components share one
//! sample clock (an accelerometer's x/y/z axes). This example also shows
//! recognized milli-g values and a custom display-only unit, then reads the
//! file back and prints its channel metadata.

// Examples favour brevity: the core's robustness restriction lints (no
// unwrap/expect/panic/indexing, no std::env::temp_dir) are not enforced on
// demonstration code, mirroring how clippy.toml relaxes them inside tests.
#![allow(
    clippy::restriction,
    clippy::cognitive_complexity,
    clippy::disallowed_methods,
    clippy::allow_attributes,
    reason = "SDK example: demonstration code"
)]

use std::{env, error::Error, fs};

use geotrace_sdk::{
    Angle, Channel, ChannelUnit, DateTime, Duration, NavFile, NavFileBuilder, NavFix, NavFixTime,
    Unit, Utc,
};

fn main() -> Result<(), Box<dyn Error>> {
    let t0 = "2024-06-01T08:00:00Z".parse::<DateTime<Utc>>()?;

    let mut recorder = NavFileBuilder::new().with_title("Channel tour").open();
    recorder.add(
        NavFix::builder()
            .time(NavFixTime::Receiver(t0))
            .lat(Angle::degrees(51.5074))
            .lon(Angle::degrees(-0.1278))
            .build(),
    );

    // Three samples, one second apart. A real recorder would sample faster
    // than the fixes. The channel keeps its own clock either way.
    let times: Vec<DateTime<Utc>> = (0..3).map(|i| t0 + Duration::seconds(i)).collect();

    // A scalar channel: one value per timestamp.
    recorder.add(
        Channel::builder()
            .name("incline")
            .unit(Unit::DEG)
            .description("boom inclinometer")
            .times(times.clone())
            .values(vec![1.0, 1.5, 2.0])
            .build()?,
    );

    // A vector channel: `values` is row-major, one row of x/y/z per
    // timestamp. Declaring the unit as `mg` lets GeoTrace's query
    // language compare it against acceleration literals.
    recorder.add(
        Channel::builder()
            .name("accel")
            .unit(Unit::MG)
            .description("IMU acceleration")
            .components(["x", "y", "z"])
            .times(times.clone())
            .values(vec![
                0.0, 200.0, 980.0, //
                100.0, 200.0, 980.0, //
                200.0, 200.0, 980.0,
            ])
            .build()?,
    );

    // A custom unit is displayed verbatim and remains dimensionless in queries.
    recorder.add(
        Channel::builder()
            .name("quality")
            .unit(ChannelUnit::custom("vendor score")?)
            .times(times)
            .values(vec![80.0, 81.0, 82.0])
            .build()?,
    );

    let nav_file = recorder.finish()?;

    let path = env::temp_dir().join("geotrace_channels.gtd");
    nav_file.write_to_file(&path)?;

    let loaded = NavFile::open(&path)?;
    println!("{} channels:", loaded.channels().len());
    for channel in loaded.channels() {
        print!("  {:<10} {} samples", channel.name(), channel.times().len());
        if let Some(unit) = channel.unit() {
            print!(" [{unit}]");
        }
        if !channel.components().is_empty() {
            print!(" components: {}", channel.components().join(" "));
        }
        println!();
    }

    fs::remove_file(&path)?;
    Ok(())
}
